quinn-proto/Cargo.toml

```diff
@@ -36,6 +36,10 @@ tinyvec = { version = "1.1", features = ["alloc"] }
 tracing = "0.1.10"
 webpki = { version = "0.22", default-features = false, optional = true }
 
+[target.'cfg(all(target_arch = "wasm32", target_os = "unknown"))'.dependencies]
+# Choose appropriate feature set for getrandom, which is used via rand/rand_core
+getrandom = { version = "0.2", features = ["js"] }
```

Comment:
Are we sure this doesn't affect other targets? My understanding is that target-specific features don't actually work yet (https://github.com/rust-lang/cargo/issues/1197).

Reply:
Good call-out - this one always trips me up so I've gone back for a closer look. I'm pretty sure this is okay. The deficiency in cargo is that if your dependency appears multiple times in `Cargo.toml` underneath different targets, you'll end up with the union of features rather than the specific set you wanted for your target. Choosing whether the dependency exists or not (with a given featureset) does work though.

As committed:
* `cargo tree` (for x86_64) shows no direct dependency on `getrandom` from `quinn-proto` (but there is one under `rand_core` of course, which doesn't have any JS support libs)
* `cargo tree --target wasm32-unknown-unknown` has a direct dependency, and has also brought in the JS-related libs

Interestingly, it also works fine if we take out the `[target...]` header entirely and leave `getrandom`+`js` as an always-on dependency.

* `cargo tree` (for x86_64) has the direct dependency on `getrandom` but it has no meaningful effect, and _doesn't_ pull in JS-related dependencies
* `cargo tree --target wasm32-unknown-unknown` is the same as before, and works as you would expect
* `getrandom` itself has target-specific dependencies for that extra JS support, which is why the feature is not harmful on a more typical platform

https://github.com/rust-random/getrandom/blob/7089766df0ae5b9377d5bf63138fad235823ea6d/Cargo.toml#L26-L28

So we could just pop it straight in `[dependencies]` if this target pruning is just going to create noise/confusion.

(For reference, this "add `getrandom` to your own `Cargo.toml`" is what they recommend [in their docs](https://docs.rs/getrandom/latest/getrandom/index.html#indirect-dependencies) to support WASM builds.)

Reply:
Thanks for the detailed investigation! It looks like `getrandom` is taking care to make this safe. In particular:

> This feature has no effect on targets other than `wasm32-unknown-unknown`.

is reassuring. On that basis, let's preserve this as currently written; might as well avoid spurious dependency edges.

Reply:
Although... any idea why this clause doesn't exist in `rand` or `rand_core` itself? It seems odd for it to be specifically our responsibility to apply this target detection.

Reply:
Tbh I had the same thought but the the `rand` issues (https://github.com/rust-random/rand/issues/1110#issuecomment-812909245 https://github.com/rust-random/rand/issues/991#issuecomment-771386545) make it clear that they don't want to do feature-forwarding.

> We intentionally have removed such feature-forwarding and we do not recommend addition of such features in library crates dependent on `getrandom`.

Reply:
Actually, now that I think a bit harder it means that I shouldn't be putting this in `quinn-proto` but rather in the project that depends on `quinn-proto`. I'll take it out!

Reply:
I mean, I'm confused as to why they don't just enable the feature themselves. It sounds like @newpavlov is against *any* library crates enabling this feature, which suggests that we shouldn't either, though I don't understand the reasoning.

Reply:
The idea is that the `js` feature SHOULD be enabled only by application crates. Enabling it by default will make life miserable for those who target `wasm32-unknown-unknown`, but do not rely on `wasm-bindgen`. Ideally we would have a separate target for "web" WASM. Alternatively `js` could've been a configuration flag, but unfortunately those are somewhat unergonomic to use.

Reply:
Thank you for clarifying! I'm glad we were able to dodge that bullet at review time.